### PR TITLE
Echelon system

### PR DESCRIPTION
**Pull request by alexkar598**
# Document the changes in your pull request

Includes the changes of #18859 

Nothing major in here

Ready for review

Am paid by xanny to do this

### CODE REVIEW

File: .github/workflows/turdis.yml [hunk 1]
```diff
@@ -133,8 +133,6 @@ jobs:
           #tools/travis/install_extools.sh
           mysql -u root -h 127.0.0.1 -e 'CREATE DATABASE tg_travis;'
           mysql -u root -h 127.0.0.1 tg_travis < SQL/tgstation_schema.sql
-          mysql -u root -h 127.0.0.1 -e 'CREATE DATABASE tg_travis_prefixed;'
```
Comment (adamsong) [inline .github/workflows/turdis.yml:136]: Atomize this

File: SQL/database_changelog.txt [hunk 2]
```diff
@@ -1,13 +1,31 @@
...
+Adds allow_vpn to bound credentials flags
+
+ALTER TABLE `bound_credentials` MODIFY COLUMN flags set('bypass_bans','allow_proxies') DEFAULT NULL NULL;
+CREATE TABLE `proxy_cache` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `ip` int(11) unsigned NOT NULL,
+  `data` mediumtext NOT NULL,
+  `last_updated` timestamp NOT NULL DEFAULT current_timestamp() ON UPDATE current_timestamp(),
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `ip` (`ip`)
```
Comment (adamsong) [inline SQL/database_changelog.txt]: Just make ip the primary key

File: SQL/tgstation_schema.sql [hunk 3]
```diff
@@ -3,6 +3,7 @@
 /*!50503 SET NAMES utf8mb4 */;
 /*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
 /*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
+SET GLOBAL event_scheduler=ON
```
Comment (adamsong) [inline SQL/tgstation_schema.sql]: This should likely be in the changelog

File: .github/workflows/turdis.yml (hunk 1 above)
Comment (alexkar598) [reply .github/workflows/turdis.yml:136]: See PR description, PR is to be merged after #18859

**Review by adamsong — CHANGES_REQUESTED** [review]

File: SQL/tgstation_schema.sql (hunk 3 above)
Comment (alexkar598) [reply SQL/tgstation_schema.sql]: I got rid of the set, set commands dont persist. Dunno how to describe required mariadb configs

Comment (alexkar598) [reply SQL/tgstation_schema.sql]: That being said, it does display a warning if you try to add an even without the event scheduler being turned on

File: code/__DEFINES/subsystems.dm [hunk 4]
```diff
@@ -156,6 +158,7 @@
 #define INIT_ORDER_OVERLAY			-6
 #define INIT_ORDER_XKEYSCORE		-10
 #define INIT_ORDER_STICKY_BAN		-10
+#define INIT_ORDER_ECHELON      -10
```
Comment (adamsong) [inline code/__DEFINES/subsystems.dm]: Alignment

**Review by adamsong — CHANGES_REQUESTED** [review]

File: SQL/database_changelog.txt (hunk 2 above)
Comment (alexkar598) [reply SQL/database_changelog.txt]: Done

File: code/__DEFINES/subsystems.dm (hunk 4 above)
Comment (alexkar598) [reply code/__DEFINES/subsystems.dm]: Fixed, ish

**Review by adamsong — APPROVED** [review]
lgtm, lets see how the tm goes

File: code/modules/admin/IsBanned.dm [hunk 5]
```diff
@@ -137,6 +137,11 @@ Yogs End*/
 				
 				qdel(query_get_bound_creds)
 
+				if(SSechelon.is_match(ckey, address))
+					log_access("Failed Login: [key] [computer_id] [address] - Blocked due to proxy")
+					key_cache[key] = FALSE
+					return list("reason" = "proxy", "desc" = "Your connection has been blocked due to using a VPN or proxy service. If this was done in error or if you are required to use a VPN to connect to the server, please visit [CONFIG_GET(string/banappeals) || "the forums"].")
```
Comment (adamsong) [inline code/modules/admin/IsBanned.dm]: Clarify that disabling the vpn is an option

**Review by adamsong — CHANGES_REQUESTED** [review]

File: code/modules/admin/IsBanned.dm (hunk 5 above)
Comment (alexkar598) [reply code/modules/admin/IsBanned.dm]: Done

**Review by adamsong — APPROVED** [review]
lgtm

**Review by ToasterBiome — APPROVED** [review]

**Pull request merged**